ehrql/backends/tpp.py

```diff
@@ -109,9 +109,24 @@ def modify_query_variables(self, variables):
         # return it unmodified
         if self.include_t1oo:
             return variables
-        # Otherwise we add an extra condition to the population definition which is that
-        # the patient does *not* appear in the T1OO table.
-        extra_condition = ~t1oo.exists_for_patient()
+        # Otherwise we add an extra condition to the population definition to include
+        # only patients who we know have not opted out
+        patients = ehrql.tables.tpp.patients
+        registrations = ehrql.tables.tpp.practice_registrations
+        extra_condition = (
+            # Include only patients who do *not* have a T1OO recorded
+            ~t1oo.exists_for_patient()
+            & (
+                # Of those, include only those who have a currently active practice
+                # registration (otherwise they may have recorded a T1OO elsewhere) ...
+                registrations.where(
+                    registrations.end_date.is_null()
+                ).exists_for_patient()
+                # ... or those who have died and therefore cannot have registered a T1OO
+                # elsewhere
```

Comment:
What about a patient who has de-registered with TPP, opted out while elsewhere, and then died?  I think this is ok, because the presence of a date of death on the patients table indicates that they haven't de-registered before dying, but it'd be worth being really explicit about this.

Reply:
Yes, the comment could more precise here. It should be something like:
> ... or those whose death is recorded in TPP and therefore were registered with a SystmOne practice at time of death and cannot have gone on to register a T1OO elsewhere

Reply:
That's an improvement!